packages/jest-core/src/SearchSource.ts

```diff
@@ -37,8 +37,55 @@ export type TestSelectionConfig = {
   watch?: boolean;
 };
 
-const globsToMatcher = (globs: Array<Config.Glob>) => (path: Config.Path) =>
-  micromatch([replacePathSepForGlob(path)], globs, {dot: true}).length > 0;
+const globsMatchers = new Map<
+  string,
+  {
+    isMatch: (str: string) => boolean;
+    negated: boolean;
+  }
+>();
+
+const globsToMatcher = (globs: Array<Config.Glob>) => {
+  const matchers = globs.map(glob => {
+    if (!globsMatchers.has(glob)) {
+      const state = micromatch.scan(glob, {dot: true});
+      const matcher = {
+        isMatch: micromatch.matcher(glob, {dot: true}),
+        negated: state.negated,
+      };
+      globsMatchers.set(glob, matcher);
+    }
+    return globsMatchers.get(glob);
+  });
+
+  return (path: Config.Path) => {
+    const replacedPath = replacePathSepForGlob(path);
+    let omit = false;
+    let keep = false;
+    let negatives = 0;
+
+    for (let i = 0; i < matchers.length; i++) {
+      const matcher = matchers[i];
+      if (!matcher) continue;
```

Comment:
can this happen?

you can add a `!` to the `get` on line 56 as we know it's always there